sgkit/stats/aggregation.py

```diff
@@ -167,6 +185,33 @@ def count_variant_alleles(ds: Dataset, merge: bool = True) -> Dataset:
     return conditional_merge_datasets(ds, new_ds, merge)
 
 
+def count_cohort_alleles(ds: Dataset, merge: bool = True) -> Dataset:
+
+    n_variants = ds.dims["variants"]
+    n_alleles = ds.dims["alleles"]
+
+    ds = count_call_alleles(ds)
+    AC, SC = da.asarray(ds.call_allele_count), da.asarray(ds.sample_cohort)
+    n_cohorts = SC.max().compute() + 1  # 0-based indexing
+    C = da.empty(n_cohorts, dtype=np.uint8)
+
+    G = da.asarray(ds["call_genotype"])
+    shape = (G.chunks[0], n_cohorts, n_alleles)
+
+    AC = da.map_blocks(_count_cohort_alleles, AC, SC, C, chunks=shape, dtype=np.int32)
+    assert_array_shape(
+        AC, n_variants, n_cohorts * AC.numblocks[0], n_alleles * AC.numblocks[1]
+    )
+
+    # Stack the blocks and sum across them
+    # (which will only work because each chunk is guaranteed to have same size)
+    AC = da.stack([AC.blocks[:, i] for i in range(AC.numblocks[1])]).sum(axis=0)
+    assert_array_shape(AC, n_variants, n_cohorts, n_alleles)
+
+    new_ds = Dataset({"cohort_allele_count": (("variants", "cohorts", "alleles"), AC)})
+    return conditional_merge_datasets(ds, new_ds, merge)
```

Comment:
@tomwhite A quick question, I am trying to understand what this function (`count_cohort_alleles`) is trying to do mathematically, is there a reference implementation or formula, which I can look at to understand this better? Or is it just a an attempt to do a better rewrite of an old implementation?

Reply:
@aktech Good question (and this function definitely needs documentation!). It's needed because `count_call_alleles` and `count_variant_alleles` return allele counts for each sample and for all samples (respectively), whereas in this PR we need something between the two since we now have the concept of "cohort" (a set of samples): so we need per-cohort counts.

The counts array computed here is of shape `(n_variants, n_cohorts, n_alleles)`, so we have a count for each variant/cohort/allele combination. Have a look at the tests (e.g. `test_count_cohort_alleles__multi_variant_multi_sample`), and perhaps try calling the allele counting functions in a Python repl to get a feel for what they are doing.